lms/validation/_launch_params.py

```diff
@@ -0,0 +1,80 @@
+"""Schema for validating LTI launch params."""
+import marshmallow
+from webargs.pyramidparser import parser
+from pyramid.httpexceptions import HTTPUnprocessableEntity
+
+from lms.services import LTILaunchVerificationError
+from lms.validation._exceptions import ValidationError
+from lms.values import LTIUser
+
+
+__all__ = ("LaunchParamsSchema",)
+
+
+class LaunchParamsSchema(marshmallow.Schema):
+    """
+    Schema for LTI launch params.
+
+    Validates and verifies LTI launch params. Usage via webargs::
+
+        >>> from webargs.pyramidparser import parser
+        >>>
+        >>> schema = LaunchParamsSchema(request)
+        >>> parsed_params = parser.parse(schema, request, locations=["form"])
+
+    Or to verify the request and get an :class:`~lms.values.LTIUser`
+    from the request's params::
+
+        >>> schema = LaunchParamsSchema(request)
+        >>> schema.lti_user()
+        LTIUser(user_id='...', ...)
+    """
+
+    user_id = marshmallow.fields.Str(required=True)
+
+    oauth_consumer_key = marshmallow.fields.Str(required=True)
+    oauth_nonce = marshmallow.fields.Str(required=True)
+    oauth_signature = marshmallow.fields.Str(required=True)
+    oauth_signature_method = marshmallow.fields.Str(required=True)
+    oauth_timestamp = marshmallow.fields.Str(required=True)
+    oauth_version = marshmallow.fields.Str(required=True)
+
+    class Meta:
+        """Marshmallow options for this schema."""
+
+        # Silence a strict=False deprecation warning from marshmallow.
+        # TODO: Remove this once we've upgraded to marshmallow 3.
+        strict = True
+
+    def __init__(self, request):
```

Comment:
Hmm. The fact that the schema depends on the request and services associated with it gives me a slight pause ... but we're not envisioning using this schema outside of a request context at present.

Reply:
I think it's fine. The launch verifier service in turn depends on config settings and on the URL, method, headers and parameters of the request. You can't really do the validation and verification that's being done here without a request. I think it might be desirable to be able to _construct_ a `LaunchParamsSchema` without a request (and then pass the request object in as an argument to the validation methods) but we'll see if that comes up in future.